### ISSUE TITLE
Add coalesce argument to join

### ISSUE THREAD

**Issue by orlp** [pola-rs/polars#14925]
### Description

There has been some discussion recently on whether or not we should coalesce the key columns in a join or not, and we've added `outer_coalesce` and were planning to add `left_coalesce`.

I propose we instead add a new argument to [`join`](https://docs.pola.rs/py-polars/html/reference/dataframe/api/polars.DataFrame.join.html) called `coalesce`, with default value `None`.

If `coalesce = False`, then we always output both the left and the right key column, except `semi` and `anti`. Note that this is a bit silly for `inner`, because in the first two cases the left and right column are always equal but nevertheless, it is a valid result.

If `coalesce = True`, then we coalesce the left and right key column, with the name of the left key column.

I'd suggest that the default `coalesce = None` has the effect of `True` for `inner`, and `False` for the rest. This ensures that by default nothing changes for current users, and I think it makes the most sense in general. Finally I suggest we deprecate `outer_coalesce` in favor of the option.

This supercedes https://github.com/pola-rs/polars/issues/13441.

---

To summarize what would change, the default would be the same as now, and:

 - for `inner` a *new* column would appear that is equal to the left key column with the name of the right key column,
 - for `semi` and `anti` nothing changes regardless of the parameter,
 - for the other joins when `coalesce = True` the right key column would disappear from the output dataframe and its values would appear in the left key column when the corresponding left value would normally be null.

**Comment by gab23r** [comment]
I like the idea of adding a new argument, however I disagree with the default behavior. I would be in favor of `coalesce = False` being the default. Few reasons:
- Speed: a constant in polars is that using the default argument leads to speed up. Materializing another column by default will slow down the pipeline down the road (imagine a filter after the join in eager mode).
- Convenience: in EDA this will lead to plenty of `..._right` columns all over the place, things get really messy quicky (even worse: if you join two times on the same column say id, this will throw an error saying id_right is duplicated...)
- Simplicity: with `coalesce = False` the behavior is not depending on another argument which make it really easy to reason about. 
- Back compatibility: With this default, we have a breaking change only for the `outer_coalesce` case. For `left` and and `outer` (IMO 95% of the cases) this isn't a breaking change.
